### DISCUSSION TITLE
Cell tracking using CellPose in the segementation module

### DISCUSSION THREAD

**Discussion by ML33M** [SchmollerLab/Cell_ACDC#524] [Q&A] [answered]
Hi, I have a question regarding cell tracking using the segmentation model. 
Background: After manually sorting out the test images into correct data folder structures using the module 0, I can launch the segmentation module to employ default cellpose v2 cyto for segmentation of my bright field image (Time point1, sits in the position 1 folder). It worked quite well, and I really enjoy the visual and object label. 

Now I would like to expand this to the other 2 positions, and follow the manual, I can load the entire experiment folder with all 3 positions into ACDC and  the model selection tool pop up, now I'm not sure how to use it for 1) segment each image(position), and 1) track the segmented objects from position 1 to position 3. By that I mean I would like to somehow know the object 1 from position 1 is the same object 1 in position 2 and 3, so on so forth, so later I can visually inspect the 3 positions, and export a csv of their location and measurements if possible.

**Comment by ElpadoCan** [comment]
Hi @ML33M, I'm happy to hear you are starting to enjoy Cell-ACDC! 

For tracking across timepoints, each video needs to be in a single Position, i.e., you need to merge the timepoints from different Positions into a single TIFF file. Then you can load the entire video and track the objects in it. 

To achieve this, you can use module 0 where you click on "Re-structure image files", and then you select `Multiple files, one for each time-point`. However, you will first need to put all the timepoints files into a single folder. Also, the filenames should end with something like `t01.tif`, `t02.tif`, for each timepoint. 

Once you have the entire video into a single Position folder, you will have access to all the tracking tools in Cell-ACDC.

Let me know if it works or if you have issues, thanks!

**Reply by ML33M** [reply]
Hi @ElpadoCan thank you for the reply, and thank you guys for making this tool. I'm testing it and trying to propagate the knowledge to my lab and others in Australia.  I really love the flexibility to have DL models and upload custom models. Back to my question: I see how I stuffed up, I chose the "multiple files, one for each channel". I think I shot myself in the foot when editing the names of the raw .tiff images by adding the frames number before channel name string. I tested your way and it did work. 

Now on to the fun part. So technically they are not videos, they are just snap shots of the same well taken days apart (we are a screening lab so no fancy microscopy or video camera). However, I have images before and 48 hrs after drug treatment, the spheroid did move a bit (compounded by researcher taking photos at different focal height). Nevertheless, the ACDC tracker worked very decently, I can see plenty cells dropped out from frame 1 to 2 ,but some are retained (threshold at 0.5). I have 3 basic questions: a) regarding different tracker, is there anywhere I can read about them? Like simple documentations rather than research papers? b) based on the saved data output; 2) from the GUI when I loaded the segmentation results, from frame 1 to 2, when I turn on the IDs, I see 3 colors, am I correct the yellow ones are the "new" objects, "red" are the ones lost in tracking, and "white" labels are the IDs pass the threshold and being tracked?  3) from the output file, am I safe to assume the same number in the "Cell_ID" column from different frames are the same objects being tracked?

**Reply by ElpadoCan** [reply]
> I think I shot myself in the foot when editing the names of the raw .tiff images by adding the frames number before channel name string. I tested your way and it did work.

No that's fine, the reason why we have that utility is because you are not the first user to have separated tiff files. However, previous users had the timepoint information at the end of the filename. There is no other specific reason and I could have added another template matching, that is pretty easy. In any case, it sounds like you solved this issue right?

> regarding different tracker, is there anywhere I can read about them? Like simple documentations rather than research papers?

Unfortunately not, also because most of the trackers are from other research groups. You gave me a cool idea anyway and I will try to come up with some documentation soon. In the meantime, if you have specific questions, feel free to ask. 

> 2) from the GUI when I loaded the segmentation results, from frame 1 to 2, when I turn on the IDs, I see 3 colors, am I correct the yellow ones are the "new" objects, "red" are the ones lost in tracking, and "white" labels are the IDs pass the threshold and being tracked?

Yellow contours with and ID and a question mark are the lost IDs, i.e., objects whose ID was not found in current frame. Red are the new IDs. If tracking failed, it is likely that a new object is not new but it should have the ID of a yellow contour object. White labels instead are the IDs that were tracked from one frame to another (however it doesn't mean tracking was 100% correct, it could be wrong there too). 

> 3) from the output file, am I safe to assume the same number in the "Cell_ID" column from different frames are the same objects being tracked?

Yes, absolutely. Cell-ACDC should not recycle IDs but I don't bet my life on it, in any case, an ID at a specific `frame_i` is absolutely unique and it is the tracked ID. 

I hope I answered all questions, if not feel free to ask.

**Reply by ML33M** [reply]
@ElpadoCan  thank you. I think the documentation or even just general rule of thumb for different tracking and segmentation would be great (even just to add more of them to the ? button next to parameters). I guess once we have more user cases, you guys can start a youtube repo on the how -to.

**Reply by ElpadoCan** [reply]
Absolutely, video tutorials have been asked by many users and it is definitely on the to-do list! I will keep you updated when I write something about the trackers, have a good day!

**Comment by ElpadoCan** [comment]
By the way, I was just thinking that to structure the files it might be worth starting from the raw microscopy files. What microscope file do you have? Do you have multiple videos in the same file? Are you on Windows or Mac?

**Reply by ML33M** [reply]
@ElpadoCan yes you are right, that's part of the reason why I was naughty and did not go for bioformat, and I manually picked a few images from the raw .tif. We have a couple high content systems in the house, they all generate .tif files except the Thermofisher ones (not too hard to convert).  In terms of "videos", unfortunately we do not have real videos, the best I can do is like 30 mins intervals. And in reality most likely bright field images taken at different dates. It is a challenge when it comes to "days": too many things moving too far, even in 3D matrices. but I think using 30 mins intervals, I can piece together a pretty decent tracking (also playing the trick of not seeding too many cells). Myself is on a Mac, my work PC allocated has 4 core and 16 gb RAM that is struggle to open a large powerpoint.... But that is just $ problem I need to convince my boss later. I'm trying to build some pipelines so we can apply ACDC to track high thoughput screens. So I'm really curious to see next step where I have CAR-T and cancer spheroid time points, how would the tracker behave, and feel free to let me know if there is a way to set up tracker to only track things of certain size (aka, small object filter to track CAR-T, and large object to track Targets)

**Reply by ElpadoCan** [reply, marked as answer]
> feel free to let me know if there is a way to set up tracker to only track things of certain size (aka, small object filter to track CAR-T, and large object to track Targets)

Sure, the way I would do it is by generating one segmentation file for the large objects and one for the small objects. You can use filters in Cell-ACDC where you filter by size. I see this more like a segmentation problem rather than a tracking one because the tracker will track whatever was segmented. I hope this gives you some things to try and potential directions! Good luck with your endeavours!
